operator: Skip reconcile for non-GAMMA xRoutes without a Cilium-managed Gateway

aetimmes opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

This is a revival of https://github.com/cilium/cilium/pull/34753.

As of 1.16.3, the Cilium Gateway API operator currently makes several assumptions around it being the only Gateway API reconciler in a given cluster; deploying the Cilium operator alongside another Gateway API operator causes an infinite back-and-forth loop of reconciliation.

This PR makes the following changes:
- adds logic to the (non-GAMMA) xRoute reconcilers to check for the existence of the Cilium gateway operator class
- ~~adds a test to ensure that the HTTPRoute reconciler does not update HTTPRoutes that do not have a cilium-managed Gateway in its parentRefs~~
- ~~updates the `non-existent gateway` test to ensure that the operator correctly errors without updating the status when a malformed HTTPRoute is encountered~~

(The above-mentioned tests have been removed in the most current iteration of this PR since the logic has been moved out of the reconciliation functions and into the builder predicates.)

Fixes: https://github.com/cilium/cilium/issues/34754

```release-note
operator: don't reconcile non-GAMMA xRoutes without a Cilium-managed Gateway
```

aetimmes commented on operator/pkg/gateway-api/helpers.go:
```diff
@@ -178,3 +179,22 @@ func setMergedLabelsAndAnnotations(temp, desired client.Object) {
+// Returns true if at least one parent gateway is managed by Cilium, false otherwise.
+func hasCiliumManagedGateway(ctx context.Context, c client.Client, parentRefs []gatewayv1.ParentReference, getGatewayFn func(ref gatewayv1.ParentReference) (*gatewayv1.Gateway, error), logger *slog.Logger) (bool, error) {
+	for _, parent := range parentRefs {
+		if (parent.Group == nil || *parent.Group == gatewayv1.GroupName) &&
```
This check exists in the helpers module, but when I tried to import/reference that method here I got some mod-vendor compilation errors that I didn't have enough golang-fu to resolve, so I've inlined them for now.

aetimmes commented on operator/pkg/gateway-api/helpers.go:
```diff
@@ -178,3 +179,22 @@ func setMergedLabelsAndAnnotations(temp, desired client.Object) {
+
+			gw, err := getGatewayFn(parent)
+			if err != nil {
+				return false, fmt.Errorf("failed to get Gateway: %w", err)
```
@sayboras The end result of this line is that if there is no Gateway associated with a route, errors will be logged - I don't know to what extent this behavior is breaking existing integration tests and whether this particular failure mode should be a warning instead of an error.

sayboras replied:
sorry for late reply, I am taking a look at this PR and get back to you soon.

sayboras replied:
WDYT about expanding the predicate function ? 

https://github.com/cilium/cilium/blob/ac46da9fcc48dce1d5f9e7ca4d600aa8332622b1/operator/pkg/gateway-api/httproute.go#L131

aetimmes replied:
Just pushed a new set of commits that adopts that approach; will test locally and see if this is viable

aetimmes replied:
Looks like this method is working as intended:

`cilium-operator-86649d6bc8-xlk9r cilium-operator 2024-12-28T17:00:46.774617211-05:00 time=2024-12-28T22:00:46Z level=debug msg="Gateway does not have matching controller, skipping" module=operator.operator-controlplane.leader-lifecycle.gateway-api controller=gateway resource=default/non-cilium-httproute`

sayboras replied:
thanks for your help investigating and confirming the result :heavy_check_mark:

aetimmes committed:
cilium-operator: don't watch non-Cilium xRoutes

Fixes: #34754

Currently, the cilium-operator xRoute reconcilers reconcile all
(non-GAMMA) xRoutes regardless of their parentage. This can lead to
race conditions where the Cilium operator can clobber metadata updates
made by other operatores (ie, the Kong Kubernetes Ingress Controller).
This PR adjusts the xRoute reconcilers to skip routes associated with the Cilium gateway.

Signed-off-by: Andrew E. Timmes <atimmes@seatgeek.com>

aetimmes commented on operator/pkg/gateway-api/grpcroute.go:
```diff
@@ -62,7 +62,8 @@ func (r *grpcRouteReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	builder := ctrl.NewControllerManagedBy(mgr).
 		// Watch for changes to GRPCRoute
-		For(&gatewayv1.GRPCRoute{}).
+		For(&gatewayv1.GRPCRoute{},
+			builder.WithPredicates(predicate.NewPredicateFuncs(r.hasMatchingGatewayParent()))).
```
With the hasMatchingGatewayParent logic moving into the builder predicates (as opposed to being inside the reconciler functions in previous iterations of this MR), I'm no longer sure where to put tests for this condition. Would be open to ideas here - maybe writing a quick+dirty unit test for the `hasMatchingGatewayParent` function would be easiest?

sayboras replied:
sounds good to me :heavy_check_mark:

aetimmes replied:
Just pushed `operator/pkg/gateway-api/httproute_test.go` with those tests. Let me know if that's the right place or if we want to merge them into the reconciler test file

sayboras reviewed (approved):
Thanks a lot for your continued work on this issue :+1:
